sandbox/mooncake.py

```diff
@@ -0,0 +1,252 @@
+        self.conv2 = nn.Conv1d(num_channels, 2 * num_channels, kernel_size=1)
+
+        self.blocks = nn.ModuleList([
+            ResidualBlock(
+                num_channels, kernel_size=kernel_size,
+                dilation=dilations[i]).to(DEVICE)
+            for i in range(self.num_blocks)
+        ])
+
+        self.conv_pi = nn.Conv1d(
+            2 * num_channels, num_components + 1, kernel_size=1, bias=False)
+        self.conv_mu = nn.Conv1d(
+            2 * num_channels, num_components, kernel_size=1, bias=False)
+        self.conv_sigma = nn.Conv1d(
+            2 * num_channels, num_components, kernel_size=1, bias=False)
+
+    def forward(self, x):
+        """
+        Parameters
+        ----------
+        x : [batch_size, num_samples]
+            Input.
+
+        Returns
+        -------
+        pi, mu, sigma : [batch_size, num_components (+1 for pi), num_samples]
+            Mixture weights, means and standard deviations of each Gaussian
+            component, respectively.
+        """
+        x = x.unsqueeze(1)  # Sinewave only has 1 channel
+        x = self.conv1(x)
+
+        residual = x
+        for i in range(self.num_blocks):
+            residual = self.blocks[i](residual)
+            x += residual
+
+        z = F.relu(x)
+        z = self.conv2(z)
+
+        pi = F.softmax(self.conv_pi(z), dim=1)
+        mu = torch.tanh(self.conv_mu(z))
+        sigma = F.softplus(self.conv_sigma(z))
+
+        return pi, mu, sigma
+
+
+if __name__ == "__main__":
+    torch.set_default_tensor_type(torch.FloatTensor)
+    np.random.seed(1618)
+    torch.manual_seed(1618)
+
+    batch_size = 4096
+    max_num_samples = 200
+
+    mooncake = Mooncake().to(DEVICE)
+    optimizer = optim.Adam(mooncake.parameters(), lr=0.001)
+
+    # Train
+    mooncake.train()
+
+    for i in range(NUM_TRAIN_ITERATIONS):
+        # Generate one sinewave and make a batch out of it.
+        sinewaves = generate_sinewaves(1, max_num_samples=max_num_samples)
+        pi, mu, sigma = mooncake(sinewaves)
+
+        # Update
+        optimizer.zero_grad()
+        nlogp = negative_log_prob(sinewaves, pi, mu, sigma).sum(dim=1).mean()
+        nlogp.backward()
+        optimizer.step()
+
+        if i % 100 == 0:
+            print("[{}] nlogp: {}".format(i,
+                                          nlogp.cpu().detach().numpy().item()))
+        del nlogp
+
+    # Infer
+    mooncake.eval()
+
+    with torch.no_grad():
+        ground_truth = generate_sinewaves(1, max_num_samples=max_num_samples)
+        inferred = torch.zeros_like(ground_truth)
+        inferred[..., :SEED_LENGTH] = ground_truth[..., :SEED_LENGTH]
+
+        for i in range(SEED_LENGTH, inferred.shape[1]):
+            pi, mu, sigma = mooncake(inferred[..., :i])
+
+            pi = pi[:, 1:, -1]
+            mu = mu[..., -1]
+            sigma = sigma[..., -1]
+            inferred[..., i] = (pi * mu).sum(dim=1)
```

Comment:
fine for now. but keep in mind we might not always be inferring the mean of the gmm